Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1381 @@
+
+```happy
+  | '-' exp                  { ENeg (L.rtRange $1 <-> info $2) $2 }
+```
+
+This will also allow us to parse interesting cases, such as `-if True then 1 else 2` or `-succ 2`.
+Try it in GHCi, both cases are valid Haskell code, and we also make it valid MiniML code now.
+
+#### Parsing binary operators
+
+Now let's move on with parsing binary operators, where Happy will make us unhappy again.
+
+First, add new productions to parse each existing operator to `exp`:
+
+```happy
+  -- Arithmetic operators
+  | exp '+'  exp             { EBinOp (info $1 <-> info $3) $1 (Plus (L.rtRange $2)) $3 }
+  | exp '-'  exp             { EBinOp (info $1 <-> info $3) $1 (Minus (L.rtRange $2)) $3 }
+  | exp '*'  exp             { EBinOp (info $1 <-> info $3) $1 (Times (L.rtRange $2)) $3 }
+  | exp '/'  exp             { EBinOp (info $1 <-> info $3) $1 (Divide (L.rtRange $2)) $3 }
+  -- Comparison operators
+  | exp '='  exp             { EBinOp (info $1 <-> info $3) $1 (Eq (L.rtRange $2)) $3 }
+  | exp '<>' exp             { EBinOp (info $1 <-> info $3) $1 (Neq (L.rtRange $2)) $3 }
+  | exp '<'  exp             { EBinOp (info $1 <-> info $3) $1 (Lt (L.rtRange $2)) $3 }
+  | exp '<=' exp             { EBinOp (info $1 <-> info $3) $1 (Le (L.rtRange $2)) $3 }
+  | exp '>'  exp             { EBinOp (info $1 <-> info $3) $1 (Gt (L.rtRange $2)) $3 }
+  | exp '>=' exp             { EBinOp (info $1 <-> info $3) $1 (Ge (L.rtRange $2)) $3 }
+  -- Logical operators
+  | exp '&'  exp             { EBinOp (info $1 <-> info $3) $1 (And (L.rtRange $2)) $3 }
+  | exp '|'  exp             { EBinOp (info $1 <-> info $3) $1 (Or (L.rtRange $2)) $3 }
+```
+
+n.b.: Do NOT extract these operators to a new production and change it to `exp operator exp` because of a limitation in Happy. This will cause conflicts that will be pretty difficult to resolve.
+You could work around it using `%prec`, but you will have to list all lookahead terminals in the operator precedence table.
+
+Remember that we previously defined the precedences and associativities of each operator; otherwise, Happy would generate 168 shift/reduce conflicts. I will copy and paste them here yet again in case you've missed them:
+
+```happy
+%right '->'
+%left '|'
+%left '&'
+%nonassoc '=' '<>' '<' '>' '<=' '>='
+%left '+' '-'
+%left '*' '/'
+```
+
+With it, we now have 12 shift/reduce conflicts.
+
+You could observe that the left-hand side of an operator is always an `atom` and change it accordingly, but this will be no good.
+Even though it would remove all of the shift/reduce conflicts, it would also have the side-effect that **the tree would always be right-biased**.
+You could live with it and properly balance the tree afterward (see the "rotation" method for `Dynamic infix operators with Alex and Happy` at the end of the article on how to do it). Still, in this case, it's possible to resolve this only by tweaking the grammar, so let's do it.
```

Comment:
Would it perhaps be better to just link it?

Reply:
Done.